pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,354 @@
+}
+
+// VolumeSnapshotSource represents the source of a snapshot.
+// Exactly one of its members must be set.
+// Members in VolumeSnapshotSource are immutable.
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotSource members
+// will not be updated once specified.
+type VolumeSnapshotSource struct {
+	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
+	// object in the same namespace as the VolumeSnapshot object where the
+	// snapshot should be dynamically taken from.
+	// +optional
+	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
+	// object a user asks to statically bind the VolumeSnapshot object to.
+	// +optional
+	VolumeSnapshotContentName *string `json:"volumeSnapshotContentName,omitempty" protobuf:"bytes,2,opt,name=volumeSnapshotContentName"`
+}
+
+// VolumeSnapshotStatus is the status of the VolumeSnapshot
+type VolumeSnapshotStatus struct {
+	// NOTE: All fields in VolumeSnapshotStatus are informational for user references.
+	// Controllers MUST NOT rely on any fields programmatically.
+
+	// boundVolumeSnapshotContentName represents the name of the VolumeSnapshotContent
+	// object to which the VolumeSnapshot object is bound.
+	// If not specified, it indicates that the VolumeSnapshot object has not been
+	// successfully bound to a VolumeSnapshotContent object yet.
+	// +optional
+	BoundVolumeSnapshotContentName *string `json:"boundVolumeSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=boundVolumeSnapshotContentName"`
+
+	// creationTime, if not nil, represents the timestamp when the point-in-time
+	// snapshot was successfully cut on the underlying storage system.
+	// In dynamic snapshot creation case, it will be filled in upon snapshot creation.
+	// For a pre-existing snapshot, it will be filled in once the VolumeSnapshot object has
+	// been successfully bound to a VolumeSnapshotContent object and the underlying
+	// storage system has the information available.
+	// If not specified, it indicates that the creation time of the snapshot is unknown.
+	// +optional
+	CreationTime *metav1.Time `json:"creationTime,omitempty" protobuf:"bytes,2,opt,name=creationTime"`
+
+	// readyToUse indicates if a snapshot is ready to be used to restore a volume.
+	// In dynamic snapshot creation case, readyToUse will be set to true after underlying storage
+	// system has successfully finished all out-of-bound procedures to make a snapshot ready to
+	// be used to restore a volume.
+	// For a pre-existing snapshot, readyToUse will be set to the value returned
+	// from CSI "ListSnapshots" gRPC call if the matching CSI driver exists and supports.
+	// Otherwise, if there exists no CSI driver or the driver does not support "ListSnapshots",
```

Comment:
Do we allow pre-provisioned snapshot if there is no CSI driver found?  What if the DeletionPolicy is "Delete"?  How can we delete it without a CSI driver?

Reply:
updated